File: src/boj/Problem1174/줄어드는_수.java
package boj.Problem1174;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class 줄어드는_수 {
    private static List<Long> numbers;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        numbers = new ArrayList<>();
        for (int i = 0; i < 10; i++) {
            makeNumbers(i);
        }
        numbers.sort(Comparator.comparingLong(o -> o));
        if (numbers.size() < n) {
            System.out.println(-1);
        } else {
            System.out.println(numbers.get(n - 1));
        }
    }

    private static void makeNumbers(long i) {
        numbers.add(i);
        long last = i % 10;
        if (last <= 0) return;
        for (long j = last - 1; 0 <= j; j--) {
            makeNumbers(i * 10 + j);
        }
    }
}

File: src/boj/Problem1174/줄어드는 수.kt
package boj.Problem1174

private lateinit var numbers: ArrayList<Long>
fun main() = with(System.`in`.bufferedReader()) {
    numbers = ArrayList()
    val n = readLine().toInt()
    for (i in 0..9) {
        makeNumber(i.toLong())
    }
    numbers.sort()
    val answer = numbers.getOrNull(n - 1) ?: -1
    println(answer)
}

private fun makeNumber(i: Long) {
    numbers.add(i)
    val last = i % 10
    if (last <= 0) return
    for (j in last - 1 downTo 0) {
        makeNumber(i * 10 + j)
    }
}
